pkg/remote/git.go

```diff
@@ -0,0 +1,145 @@
+/*
+   Copyright 2020 Docker Compose CLI authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package remote
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+
+	"github.com/compose-spec/compose-go/cli"
+	"github.com/compose-spec/compose-go/loader"
+	"github.com/docker/compose/v2/pkg/api"
+	"github.com/moby/buildkit/util/gitutil"
+)
+
+func NewGitRemoteLoader() (loader.ResourceLoader, error) {
+	var base string
+	if cache_home, ok := os.LookupEnv("XDG_CACHE_HOME"); ok {
+		base = cache_home
+	} else {
+		home, err := os.UserHomeDir()
+		if err != nil {
+			return nil, err
+		}
+		base = filepath.Join(home, ".cache")
+	}
+	cache := filepath.Join(base, "docker-compose")
+
+	err := os.MkdirAll(cache, 0o700)
+	return gitRemoteLoader{
+		cache: cache,
+	}, err
+}
+
+type gitRemoteLoader struct {
+	cache string
+}
+
+func (g gitRemoteLoader) Accept(path string) bool {
+	_, err := gitutil.ParseGitRef(path)
+	return err == nil
+}
+
+var commitSHA = regexp.MustCompile(`^[a-f0-9]{40}$`)
+
+func (g gitRemoteLoader) Load(ctx context.Context, path string) (string, error) {
```

Comment:
Doing background Git fetches is always tricky because of credential helpers.

Take a look at https://github.com/tilt-dev/go-get/blob/295ccc6550921a9766a7294360a77a4d8704d826/get.go#L18-L55

I'd clone `os.Environ()` and then conditionally add `GIT_TERMINAL_PROMPT` / `GIT_SSH_COMMAND` to the env passed to the exec'd processes.

Reply:
If Env is nil, the new process uses the current process's environment, so we don't have to